// typescript-coverage-report/dist/lib/reporters/html.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.generate = void 0;
var react_1 = __importDefault(require("react"));
var server_1 = __importDefault(require("react-dom/server"));
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var util_1 = require("util");
var SummaryPage_1 = __importDefault(require("../../components/SummaryPage"));
var DetailPage_1 = __importDefault(require("../../components/DetailPage"));
var writeFile = util_1.promisify(fs_1["default"].writeFile);
var readFile = util_1.promisify(fs_1["default"].readFile);
var mkdir = util_1.promisify(fs_1["default"].mkdir);
var checkOrWriteFolder = function (outputDir) {
    if (!fs_1["default"].existsSync(outputDir)) {
        fs_1["default"].mkdirSync(outputDir);
    }
    return outputDir;
};
var includeAsset = function (filename) {
    var extension = path_1["default"].extname(filename);
    if (extension === ".js") {
        return "<script src=\"" + filename + "\" type=\"text/javascript\" charset=\"utf-8\"></script>";
    }
    if (extension === ".css") {
        return "<link href=\"" + filename + "\" type=\"text/css\" rel=\"stylesheet\">";
    }
    console.warn("includeAsset: couldn't recognise the extension " + extension);
    return "";
};
var includeAssets = function (assets) {
    return assets.map(includeAsset).join("\n");
};
var wrapHTMLContent = function (Component, props, options) {
    if (options === void 0) { options = {}; }
    var content = server_1["default"].renderToStaticMarkup(react_1["default"].createElement(Component, props));
    return "\n  <!DOCTYPE html>\n  <html>\n    <head>\n      <title>" + (options.title || "TypeScript coverage report") + "</title>\n      " + includeAsset("https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css") + "\n      " + (options.assets || "") + "\n    </head>\n    <body>\n    " + content + "\n    <p class=\"footer-text\">TypeScript Coverage Report generated by <a href=\"https://github.com/plantain-00/type-coverage\">type-coverage</a> and <a href=\"https://github.com/alexcanessa/typescript-coverage-report\">typescript-coverage-report</a> at " + new Date().toUTCString() + "</p>\n    </body>\n  </html>\n  ";
};
// NOTE: This generate function has side effect of creating the index.html
var generate = function (data, options) { return __awaiter(void 0, void 0, void 0, function () {
    var fileContent, _loop_1, _i, _a, _b, filename, _c, totalCount, correctCount, generatedFile;
    return __generator(this, function (_d) {
        switch (_d.label) {
            case 0:
                fileContent = wrapHTMLContent(SummaryPage_1["default"], {
                    fileCounts: data.fileCounts,
                    percentage: data.percentage,
                    total: data.total,
                    covered: data.covered,
                    uncovered: data.uncovered,
                    threshold: options.threshold
                }, {
                    assets: includeAssets([
                        "./assets/source-file.css",
                        "https://cdn.jsdelivr.net/npm/sorttable@1.0.2/sorttable.min.js"
                    ])
                });
                return [4 /*yield*/, writeFile(path_1["default"].join(checkOrWriteFolder(options.outputDir), "index.html"), fileContent)];
            case 1:
                _d.sent();
                _loop_1 = function (filename, totalCount, correctCount) {
                    var annotations, assetsFolder, sourceCode, detailContent;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, mkdir(path_1["default"].join(options.outputDir, "files", path_1["default"].dirname(filename)), {
                                    recursive: true
                                })];
                            case 1:
                                _a.sent();
                                annotations = data.anys.filter(function (_a) {
                                    var file = _a.file;
                                    return file === filename;
                                });
                                assetsFolder = path_1["default"].relative(filename, "assets");
                                return [4 /*yield*/, readFile(filename, "utf-8")];
                            case 2:
                                sourceCode = _a.sent();
                                detailContent = wrapHTMLContent(DetailPage_1["default"], {
                                    filename: filename,
                                    sourceCode: sourceCode,
                                    totalCount: totalCount,
                                    correctCount: correctCount,
                                    annotations: annotations,
                                    threshold: options.threshold
                                }, {
                                    title: path_1["default"].basename(filename),
                                    assets: includeAssets([
                                        "https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.58.2/codemirror.min.js",
                                        "https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.58.2/mode/javascript/javascript.min.js",
                                        "https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.58.2/codemirror.min.css",
                                        path_1["default"].join(assetsFolder, "source-file.js"),
                                        path_1["default"].join(assetsFolder, "source-file.css")
                                    ])
                                });
                                return [4 /*yield*/, writeFile(path_1["default"].join(options.outputDir, "files", filename + ".html"), detailContent)];
                            case 3:
                                _a.sent();
                                return [2 /*return*/];
                        }
                    });
                };
                _i = 0, _a = Array.from(data.fileCounts);
                _d.label = 2;
            case 2:
                if (!(_i < _a.length)) return [3 /*break*/, 5];
                _b = _a[_i], filename = _b[0], _c = _b[1], totalCount = _c.totalCount, correctCount = _c.correctCount;
                return [5 /*yield**/, _loop_1(filename, totalCount, correctCount)];
            case 3:
                _d.sent();
                _d.label = 4;
            case 4:
                _i++;
                return [3 /*break*/, 2];
            case 5:
                generatedFile = path_1["default"].resolve(path_1["default"].join(options.outputDir, "index.html"));
                console.log("View generated HTML Report at " + generatedFile);
                return [2 /*return*/];
        }
    });
}); };
exports.generate = generate;
