// typescript-coverage-report/dist/bin/typescript-coverage-report.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var commander_1 = require("commander");
var lib_1 = __importDefault(require("../lib"));
var path_1 = __importDefault(require("path"));
var _a = require("../../package.json"), version = _a.version, description = _a.description;
var _b = require(path_1["default"].join(process.cwd(), "/package.json")).typeCoverage, typeCoverage = _b === void 0 ? {} : _b;
var argvWithVersion = function (argvs) {
    var vPos = argvs.indexOf("-v");
    if (vPos > -1) {
        argvs[vPos] = "-V";
    }
    return argvs;
};
var _c = typeCoverage.outputDir, outputDir = _c === void 0 ? "coverage-ts" : _c, _d = typeCoverage.atLeast, atLeast = _d === void 0 ? 80 : _d, _e = typeCoverage.strict, strict = _e === void 0 ? false : _e, _f = typeCoverage.debug, debug = _f === void 0 ? false : _f, _g = typeCoverage.cache, cache = _g === void 0 ? false : _g, _h = typeCoverage.project, project = _h === void 0 ? "." : _h, _j = typeCoverage.ignoreFiles, ignoreFiles = _j === void 0 ? false : _j, _k = typeCoverage.ignoreCatch, ignoreCatch = _k === void 0 ? false : _k, _l = typeCoverage.ignoreUnread, ignoreUnread = _l === void 0 ? false : _l;
commander_1.program
    .version(version)
    .description(description)
    .option("-o, --outputDir [string]", "the output directory where to generate the report.", outputDir)
    .option("-t, --threshold [number]", "the minimum percentage of coverage required.", parseFloat, atLeast)
    .option("-s, --strict [boolean]", "run the check in strict mode.", strict)
    .option("-d, --debug [boolean]", "show debug information.", debug)
    .option("-c, --cache [boolean]", "save and reuse type check result from cache.", cache)
    .option("-p, --project [string]", 'file path to tsconfig file, eg: --project "./app/tsconfig.app.json"', project)
    .option("-i, --ignore-files [string[]]", 'ignore specified files, eg: --ignore-files "demo1/*.ts" --ignore-files "demo2/foo.ts"', ignoreFiles)
    .option("--ignore-catch [boolean]", "ignore type any for (try-)catch clause variable", ignoreCatch)
    .option("-u, --ignore-unread [boolean]", "allow writes to variables with implicit any types", ignoreUnread)
    .parse(argvWithVersion(process.argv));
var options = {
    /* camelCase keys matching "long" flags in options above */
    outputDir: commander_1.program.outputDir,
    threshold: commander_1.program.threshold,
    tsProjectFile: commander_1.program.project,
    strict: commander_1.program.strict,
    debug: commander_1.program.debug,
    cache: commander_1.program.cache,
    ignoreFiles: commander_1.program.ignoreFiles,
    ignoreCatch: commander_1.program.ignoreCatch,
    ignoreUnread: commander_1.program.ignoreUnread
};
lib_1["default"](options)
    .then(function (_a) {
    var percentage = _a.percentage;
    if (percentage < options.threshold) {
        console.error("\nThe TypeScript coverage " + percentage.toFixed(2) + "% is lower than the threshold " + options.threshold + "%");
        process.exit(2);
    }
    process.exit(0);
})["catch"](function (error) {
    console.error(error);
    process.exit(255);
});
